cmd/generate/main.go

```diff
@@ -9,19 +9,44 @@ package main
 
 import (
 	"flag"
+	"fmt"
 	"log"
 	"os"
 
 	"github.com/nginxinc/nginx-go-crossplane/internal/generator"
 )
 
+type stringSliceFlag []string
+
+func (s *stringSliceFlag) String() string {
+	return fmt.Sprintf("%v", *s)
+}
+
+func (s *stringSliceFlag) Set(value string) error {
+	*s = append(*s, value)
+	return nil
+}
+
 func main() {
 	var (
 		sourceCodePath = flag.String("src-path", "",
-			"the path of source code your want to generate support from, it can be either a file or a directory. (required)")
+			"The path of source code your want to generate support from, it can be either a file or a directory. (required)")
+		filter   stringSliceFlag
+		override stringSliceFlag
 	)
+	flag.Var(&filter, "filter",
+		"A list of strings specifying the directives to include in the output. "+
+			"If not provided, all directives in the src-path will be included.\n"+
+			"An example is: -filter directive1 -filter directive2...(optional)")
+	flag.Var(&override, "override",
```

Comment:
the override and it's microformat is bugging me a little bit; `main.go`'s job is generally to convert the user input into the data that the rest of the codebase (`generator.Generate` wants, and with `override` we're kinda doing half of that here (CLI args to a `[]string`), and then the other half is inside `Generate` to convert each string to a `map[string][]mask` entry.

an alternate design:

- make a new `type Override struct` that implements [`TextUnmarshaler`](https://pkg.go.dev/encoding#TextUnmarshaler)
- change `generator.Generate` to accept a `[]Override` - it's expecting the caller to parse the microformat from CLI args into a `Override`
- make a `type overrideSliceFlag []Override` here similar to `stringSliceFlag`, to convert the CLI args to an `[]Override`, returning any parse errors from `overrideSliceFlag.Set` so it fails early

Thoughts?

Reply:
> main.go's job is generally to convert the user input into the data that the rest of the codebase (generator.Generate wants, and with override we're kinda doing half of that here (CLI args to a []string),

Yeah.. The reason we need this is: some directives have different definition in crossplane vs source code(like `if` in OSS, we added `ngxConfExpr` to it in crossplane.  `ngxConfExpr` is not in OSS). And we want to provide users with flexibilities to handle such cases. After another `go generate` tool available, this tool is likely to use at some third-party modules unknown by crossplane, and the override should be very small(at most one or two directives I think, no override for most cases). I thought that's okay?

> an alternate design:
make a new type Override struct that implements [TextUnmarshaler](https://pkg.go.dev/encoding#TextUnmarshaler)
change generator.Generate to accept a []Override - it's expecting the caller to parse the microformat from CLI args into a Override
make a type overrideSliceFlag []Override here similar to stringSliceFlag, to convert the CLI args to an []Override, returning any parse errors from overrideSliceFlag.Set so it fails early

Yeah I think you are right it should be done at the same layer as `main.go`. Applied in latest commit

Reply:
@ryepup Do you have an example of how you would make use of the `TextUnmarshaler` interface here? I know [`flag.TextVar`](https://pkg.go.dev/flag#TextVar) accepts this type. How are you proposing he make use of the `TextUnmarshaler` interface with the need to `Set()` into the slice type? 

Or, is the purpose in implementing `TextUnmarshaler` to follow an established pattern that perhaps makes this type more usable else where?

Reply:
I was thinking something very similar to what was written in the latest update, still using `flag.Var` to accumulate a slice (`TextVar` didn't seem to handle slices), and `UnmarshalText` to parse each individual value:

```golang
func (overrides *overrideParserList) Set(value string) error {
	var parser overrideParser
	err := parser.UnmarshalText([]byte(value))
	if err != nil {
		return fmt.Errorf("invalid override %q: %w", value, err) // <-- I'd expand the error returned to include the bad value
	}

	*overrides = append(*overrides, parser)
	return nil
} 
```

Moving parse errors into the CLI parsing part of the program removes some possible errors later in the main generate code path, and that "fail fast" was my primary motivation. 

`TextUnmarshaler` seemed to fit since the CLI arg is a textual representation of an override, but could just as easily be a standalone constructor func like `ParseOverride(string) (Override, error)`.